oap-server/server-storage-plugin/storage-iotdb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/iotdb/base/IoTDBInsertRequest.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.oap.server.storage.plugin.iotdb.base;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+import org.apache.skywalking.oap.server.core.storage.StorageData;
+import org.apache.skywalking.oap.server.core.storage.StorageHashMapBuilder;
+import org.apache.skywalking.oap.server.core.storage.type.StorageDataComplexObject;
+import org.apache.skywalking.oap.server.library.client.request.InsertRequest;
+import org.apache.skywalking.oap.server.library.client.request.UpdateRequest;
+import org.apache.skywalking.oap.server.storage.plugin.iotdb.IoTDBClient;
+import org.apache.skywalking.oap.server.storage.plugin.iotdb.IoTDBIndexes;
+import org.apache.skywalking.oap.server.storage.plugin.iotdb.IoTDBTableMetaInfo;
+
+@Getter
+@Setter
+@ToString
+@Slf4j
+public class IoTDBInsertRequest implements InsertRequest, UpdateRequest {
+    private String modelName;
+    private long time;
+    private List<String> indexes;
+    private List<String> indexValues;
+    private List<String> measurements;
+    private List<TSDataType> measurementTypes;
+    private List<Object> measurementValues;
+
+    public <T extends StorageData> IoTDBInsertRequest(String modelName, long time, T storageData,
+                                                      StorageHashMapBuilder<T> storageBuilder) {
+        this.modelName = modelName;
+        this.time = time;
+        indexes = IoTDBTableMetaInfo.get(modelName).getIndexes();
+        indexValues = new ArrayList<>(indexes.size());
+        Map<String, Object> storageMap = storageBuilder.entity2Storage(storageData);
+
+        indexes.forEach(index -> {
+            if (index.equals(IoTDBIndexes.ID_IDX)) {
+                indexValues.add(storageData.id());
+            } else if (storageMap.containsKey(index)) {
+                // avoid `service_group` be "null" when inserting
+                if (index.equals(IoTDBIndexes.GROUP_IDX) && storageMap.get(index) == null) {
```

Comment:
Let’s apply this to other fields. I don’t think we want “null” string for null value in all other fields. @wu-sheng correct me if I’m wrong.

Reply:
If I don't use the null check for `Group_IDX`,  `String.valueOf(null)` in line 68 will return "null". In production, all other indexes  are xxx_id and they won't be null. But `Group_IDX` maybe null.

Reply:
If other non-indexed fields is null, we won't insert it into IoTDB server. Therefore, their query results will be null, won't be "null". Please see line 81.